classes/task/send_ical_notifications.php

```diff
@@ -0,0 +1,298 @@
+        
+        mtrace('[Zoom ical Notifications] Notifying Zoom instance with ID ' . $zoom->id);
+    
+        $zoom_event = $DB->get_record('event', array('instance' => $zoom->id, 'courseid' => $zoom->course, 'modulename' => 'zoom', 'eventtype' => 'zoom'));
+        if ($zoom_event) {
+
+            $cal_event = new \calendar_event($zoom_event); // To use moodle calendar event services.
+
+            $users = $this->zoom_get_users_to_notify($zoom->id);
+            $users = $this->zoom_filter_users($zoom, $users);
+
+            $from = format_string($SITE->shortname) . ' Event No Reply (via ' .  format_string($SITE->shortname) . ')';
+
+            $ical = null;
+            $filename = null;
+            
+            // html value to render in email body
+            $description = $cal_event->description; 
+            // Avoid filters - so we can manually add links as required       
+            $activity_url = $CFG->wwwroot . '/mod/zoom/view.php?id=';
+            $modinfo = get_fast_modinfo($zoom->course);
+            $course_modules = $modinfo->get_cms();
+            if ($course_modules) {
+                foreach ($course_modules as $course_mod) {
+                    if ($course_mod->modname == 'zoom' && $course_mod->instance == $zoom->id) {
+                        $activity_url .= $course_mod->id;
+                        break;
+                    }
+                }
+            } 
+            $description .= '<p>Links:<br>';
+            $description .= '-------<br>';
+            $description .= $activity_url;
+            $description .= '</p>';
+
+            // If no registration required - then the same join link can be used - only one ical object needed
+            if ($zoom->registration == ZOOM_REGISTRATION_OFF) {
+                $ical = $this->create_ical_object($zoom, $zoom_event, $cal_event, $description);
+                $filename = $this->serialize_attachment($ical);
+            }                 
+
+            foreach($users as $user) {
+                // If registration is required - then a unique registration link per user is required - different ical objects needed
+                if ($zoom->registration != ZOOM_REGISTRATION_OFF) {
+                    $ical = $this->create_ical_object($zoom, $zoom_event, $cal_event, $description, $user);
+                    $filename = $this->serialize_attachment($ical);
+                }
+                $email_success = email_to_user($user, $from, $zoom_event->name, $description, $description, $filename, $filename);
+                if ($email_success) {
+                    mtrace('[Zoom ical Notifications] Successfully emailed user ID ' . $user->id . ' for zoom instance with ID ' . $zoom->id);
+                } else {
+                    mtrace('[Zoom ical Notifications] A problem occurred while emailing user ID ' . $user->id . ' for zoom instance with ID ' . $zoom->id);
+                }
+            }
+    
+        } 
```

Comment:
Instead of assuming that the notification was successful in the execute() function, setting the successful execution time somewhere around here allows us more information about whether the notifications were sent successfully.
```suggestion

            // Set execution time for this cron job
            $this->set_notification_execution_time($zoom->id);
        } 
```